### PR TITLE
feat: Support for IEnumerable<T>.Contains (item IN values) with column values

### PR DESCRIPTION
**Pull request by fredvollmer**
I realize I did not open an issue prior to submitting this PR. I'd be happy to do so if that will help, but since we had already implemented this solution I wanted to get this up for review.

This PR adds support for translating `IEnumerable<T>.Contains()` to an `item IN values` where condition, where `values` is a column expression. (The IN expression that ships with EF Core does not support using a database column or subquery as the `values`.)

This allows for expressions like:

```csharp
            class SampleEntity
            {
                public string Id { get; set; }
                public string[] ArrayColumn { get; set; }
            }

            var results = db.SampleEntity
                .Where(e => e.ArrayColumn.Contains("someValue"))
                .ToListAsync();
```

I used npgsql's approach as guidance: https://github.com/npgsql/efcore.pg/blob/main/src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayTranslator.cs

I am new to EF Core and still have much to learn. Please go easy on me, but of course, any feedback or places for improvement would be much appreciated.

- [ ] Tests pass
- [ ] Appropriate changes to README are included in PR

### CODE REVIEW

**Comment by olavloite** [comment]
@fredvollmer Thanks for your contribution! Would you mind signing the CLA? Without it, it will not be possible to merge this PR.

File: Google.Cloud.EntityFrameworkCore.Spanner/Extensions/MethodInfoExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿using System.Reflection;
```
Comment (olavloite) [inline Google.Cloud.EntityFrameworkCore.Spanner/Extensions/MethodInfoExtensions.cs:1]: nit: All files should have a copyright header with the current year (so 2022 in this case)

File: Google.Cloud.EntityFrameworkCore.Spanner/Extensions/TypeExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+﻿#nullable enable
```
Comment (olavloite) [inline Google.Cloud.EntityFrameworkCore.Spanner/Extensions/TypeExtensions.cs:1]: nit, also here: Copyright header. Also, is the `nullable enable` necessary here?

File: Google.Cloud.EntityFrameworkCore.Spanner/Google.Cloud.EntityFrameworkCore.Spanner.csproj [hunk 3]
```diff
@@ -19,23 +19,23 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Google.Cloud.Spanner.Data" Version="4.2.0"/>
```
Comment (olavloite) [inline Google.Cloud.EntityFrameworkCore.Spanner/Google.Cloud.EntityFrameworkCore.Spanner.csproj:22]: Would you mind removing these changes. As far as I can tell, these are just whitespace differences.

File: Google.Cloud.EntityFrameworkCore.Spanner/Query/Internal/SpannerContainsExpression.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
+﻿#nullable enable
```
Comment (olavloite) [inline Google.Cloud.EntityFrameworkCore.Spanner/Query/Internal/SpannerContainsExpression.cs:1]: Here also: copyright header + remove `nullable enable`?

File: Google.Cloud.EntityFrameworkCore.Spanner/Query/Internal/SpannerQuerySqlGenerator.cs [hunk 5]
```diff
@@ -35,12 +45,15 @@ protected override string GetOperator(SqlBinaryExpression binaryExpression)
...
-                    && (binaryExpression.Right.TypeMapping.StoreTypeNameBase == "STRING" || binaryExpression.Right.TypeMapping.StoreTypeNameBase == "BYTES"))
+                if ((binaryExpression.Left.TypeMapping.StoreTypeNameBase == "STRING" ||
+                     binaryExpression.Left.TypeMapping.StoreTypeNameBase == "BYTES")
+                    && (binaryExpression.Right.TypeMapping.StoreTypeNameBase == "STRING" ||
+                        binaryExpression.Right.TypeMapping.StoreTypeNameBase == "BYTES"))
                 {
                     return "||";
                 }
             }
+
```
Comment (olavloite) [inline Google.Cloud.EntityFrameworkCore.Spanner/Query/Internal/SpannerQuerySqlGenerator.cs:56]: nit: remove unrelated changes (probably from a code formatter)

File: Google.Cloud.EntityFrameworkCore.Spanner/Query/Internal/SpannerQuerySqlGenerator.cs [hunk 6]
```diff
@@ -122,5 +137,54 @@ protected override Expression VisitSqlFunction(SqlFunctionExpression sqlFunction
 
             return base.VisitSqlFunction(sqlFunctionExpression);
         }
+
+        protected virtual Expression VisitContains(SpannerContainsExpression containsExpression)
```
Comment (olavloite) [inline Google.Cloud.EntityFrameworkCore.Spanner/Query/Internal/SpannerQuerySqlGenerator.cs:141]: The logic here seems to overlap to a large degree with the logic in `SpannerContainsExpression`. Could we combine the equal parts into one method?

File: Google.Cloud.EntityFrameworkCore.Spanner/Query/Internal/SpannerQuerySqlGenerator.cs [hunk 7]
```diff
@@ -122,5 +137,54 @@ protected override Expression VisitSqlFunction(SqlFunctionExpression sqlFunction
...
+                {
+                    if (!first)
+                    {
+                        Sql.Append(", ");
+                    }
+
+                    first = false;
+                    Sql.Append(constantValuesExpression.TypeMapping?.GenerateSqlLiteral(item) ??
+                               item?.ToString() ?? "NULL");
+                }
```
Comment (olavloite) [inline Google.Cloud.EntityFrameworkCore.Spanner/Query/Internal/SpannerQuerySqlGenerator.cs:166]: Could we use `String.Join(..)` for this? So something like `String.Join(", ", constantValues.Select(item => constantValuesExpression.TypeMapping?.GenerateSqlLiteral(item) ??
                               item?.ToString() ?? "NULL")`

File: Google.Cloud.EntityFrameworkCore.Spanner/Extensions/TypeExtensions.cs (hunk 2 above)
Comment (fredvollmer) [reply Google.Cloud.EntityFrameworkCore.Spanner/Extensions/TypeExtensions.cs:1]: Thanks @olavloite , will add the copyright header. As far as the `#nullable enable`, no, it is not absolutely necessary. However, it allows for greater type safety at compile time, by letting us declare reference types as nullable (e.g., `MyType?`.)

Importantly, the feature can be enabled on a file-by-file basis by adding the `#nullable enable` directive. My thought was that I should take advantage of it with this new contribution, even if it wasn't being used everywhere. But I can certainly remove it if needed.

Microsoft has good documentation here: https://learn.microsoft.com/en-us/dotnet/csharp/nullable-references

File: Google.Cloud.EntityFrameworkCore.Spanner/Query/Internal/SpannerQuerySqlGenerator.cs (hunk 7 above)
Comment (fredvollmer) [reply Google.Cloud.EntityFrameworkCore.Spanner/Query/Internal/SpannerQuerySqlGenerator.cs:166]: @olavloite The `Sql` variable is not actually a string, it's an `IRelationalCommandBuilder`.

File: Google.Cloud.EntityFrameworkCore.Spanner/Query/Internal/SpannerQuerySqlGenerator.cs (hunk 6 above)
Comment (fredvollmer) [reply Google.Cloud.EntityFrameworkCore.Spanner/Query/Internal/SpannerQuerySqlGenerator.cs:141]: I agree--I found it strange that EF Core required two separate methods that essentially do the same thing (building a SQL string). However, one uses an `ExpressionPrinter`, while the other uses an `IRelationalCommandBuilder`. Looking at the built-in commands in EFCore, as well as the implementations in the postgres provider, it seems that separating the logic is common practice.

**Comment by fredvollmer** [comment]
@olavloite Thanks for the review! I have signed the CLA, and integrated all of your code cleanup suggestions. Please let me know whether you'd still like me to remove the `#nullable `enable` directives, and also if you think I should do anything more in regards to merging the SQL generation logic.

File: Google.Cloud.EntityFrameworkCore.Spanner/Query/Internal/SpannerQuerySqlGenerator.cs (hunk 6 above)
Comment (olavloite) [reply Google.Cloud.EntityFrameworkCore.Spanner/Query/Internal/SpannerQuerySqlGenerator.cs:136]: Ack

File: Google.Cloud.EntityFrameworkCore.Spanner/Query/Internal/SpannerQuerySqlGenerator.cs (hunk 7 above)
Comment (olavloite) [reply Google.Cloud.EntityFrameworkCore.Spanner/Query/Internal/SpannerQuerySqlGenerator.cs:161]: Ack

**Review by olavloite — APPROVED** [review]
LGTM. Thanks for the contribution!
